test/e2e/suites/e2e/e2e_test.go

```diff
@@ -192,6 +198,59 @@ var _ = Describe("e2e tests", func() {
 		})
 	})
 
+	Describe("Workload cluster (multiple attached networks)", func() {
+		BeforeEach(func() {
+			// Create 2 additional networks to be attached to all cluster nodes
+			// We can't clean up these networks in a corresponding AfterEach because they will still be in use by the cluster.
+			// Instead we clean them up after the cluster has been deleted.
+
+			shared.Byf("Creating additional networks")
+
+			extraNet1, err := shared.CreateOpenStackNetwork(e2eCtx, fmt.Sprintf("%s-extraNet1", namespace.Name), "10.14.0.0/24")
+			Expect(err).NotTo(HaveOccurred())
+			postClusterCleanup = append(postClusterCleanup, func() {
+				shared.Byf("Deleting additional network %s", extraNet1.Name)
+				err := shared.DeleteOpenStackNetwork(e2eCtx, extraNet1.ID)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			extraNet2, err := shared.CreateOpenStackNetwork(e2eCtx, fmt.Sprintf("%s-extraNet2", namespace.Name), "10.14.1.0/24")
+			Expect(err).NotTo(HaveOccurred())
+			postClusterCleanup = append(postClusterCleanup, func() {
+				shared.Byf("Deleting additional network %s", extraNet2.Name)
+				err := shared.DeleteOpenStackNetwork(e2eCtx, extraNet2.ID)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			os.Setenv("CLUSTER_EXTRA_NET_1", extraNet1.ID)
+			os.Setenv("CLUSTER_EXTRA_NET_2", extraNet2.ID)
+		})
+
+		It("It should be creatable and deletable", func() {
+			shared.Byf("Creating a cluster")
+			clusterName := fmt.Sprintf("cluster-%s", namespace.Name)
+			configCluster := defaultConfigCluster(clusterName, namespace.Name)
+			configCluster.ControlPlaneMachineCount = pointer.Int64Ptr(3)
+			configCluster.WorkerMachineCount = pointer.Int64Ptr(1)
+			configCluster.Flavor = shared.FlavorMultiNetwork
+			md := createCluster(ctx, configCluster)
+
+			workerMachines := framework.GetMachinesByMachineDeployments(ctx, framework.GetMachinesByMachineDeploymentsInput{
+				Lister:            e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				ClusterName:       clusterName,
+				Namespace:         namespace.Name,
+				MachineDeployment: *md[0],
+			})
+			controlPlaneMachines := framework.GetControlPlaneMachinesByCluster(ctx, framework.GetControlPlaneMachinesByClusterInput{
+				Lister:      e2eCtx.Environment.BootstrapClusterProxy.GetClient(),
+				ClusterName: clusterName,
+				Namespace:   namespace.Name,
+			})
+			Expect(len(workerMachines)).To(Equal(1))
+			Expect(len(controlPlaneMachines)).To(Equal(3))
```

Comment:
I see that this creates a four node cluster with the machines having ports on two different networks. Did you also want to test your functions? I think you should be able to see e.g. `controlPlaneMachines[0].Status.Addresses`, if the controller is working, this should show multiple addresses, since they get assigned in reconciliation: 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/fcbf20746b7dfe3b7776c1d7420c341e1225d9cb/controllers/openstackmachine_controller.go#L334

Reply:
This was a lazy late addition, tbh. I'd previously written it as a standalone create/delete test which I expected to fail. However, it turns out that it eventually comes up because although IP selection was non-deterministic and failed when it picked the wrong one, the reconciler would just retry and eventually it picked the right one 🤷‍♂️

You are right: this test would be massively improved by testing the resulting addresses.

Reply:
I've expanded the e2e test to explicitly check ports and addresses.

Incidentally, @EmilienM you may want to do something like this in #1008? I'm iterating over openstack ports per machine, which I think it what we were trying to do in that test.